common/src/primitives/id.rs

```diff
@@ -133,6 +133,57 @@ pub trait Idable {
     fn get_id(&self) -> Id<Self::Tag>;
 }
 
+impl<T: Idable> Idable for &T {
+    type Tag = T::Tag;
+    fn get_id(&self) -> Id<Self::Tag> {
+        (*self).get_id()
+    }
+}
+
+/// An object together with its pre-calculated ID.
+///
+/// This only allows immutable access to the underlying object to prevent it from going out of sync
+/// with the ID, which is calculated for its contents. Getting an ID is nearly cost-free.
+#[derive(Clone, Copy, Eq, PartialEq, Debug)]
+pub struct WithId<T: Idable> {
```

Comment:
I was planning to create exactly this as a solution to id caching, so that we don't calculate it multiple times, especially during block processing. Thanks for going through with it.

Anyway, can you move this to its own rs file? I'm even OK with an `id` submodule.

Reply:
WIll move it.